---
title: "chs_v9_share_compiled"
output: html_document
editor_options: 
  chunk_output_type: console
---

Code for analysis performed to match CHS microbiome transplant paper, "GxGxE interactions and adaptive potential of the microbiome in Drosophila melanogaster". 


```{r}
rm(list=ls())

#libraries (download all even though not totally sure what's used)
#microbiome/visualization packages
require(ggplot2)
require(phyloseq)
require(vegan)
require(dplyr)
require(scales)
require(grid)
require(reshape2)
require(ggthemes)
require(car)
require(tibble)
require(tidyr)
require(RColorBrewer)
require(forcats)
require(ggpubr)

#development
require(survival)

require(survminer)

#mixed model packages
require(DHARMa)
require(glmmTMB)
require(lme4)
require(emmeans)

require(performance)

#color pallettes for various plots

#comparing by treatment 
treat.pal<-c("#99000d", "#cb181d", "#fb6a4a", "#fc9272", "#c6dbef", "#deebf7",
             "#084594", "#2171b5", "#fcbba1", "#fee0d2", "#4292c6", "#9ecae1")

#comparing by donor, frass, and recipients 
simplegen.pal<- c("#ef8a62","#b2182b","#fddbc7","#999999","#4d4d4d","#e0e0e0")


#comparing by full treatment (i.e. CCC, CHSC, etc) in the recipients 
redt.microemph.pal<-c("#fdae61", "#f46d43", "#74add1", "#4575b4",
                      "#d73027", "#a50026", "#313695", "#08306b")

#for fourth year talks in order of: control, HS, sterile  
microbiome.pal4<- c("#99000d","#91bfdb","#fee090")

#need to customize the dodge so you can see the difference
pd<- position_dodge(0.1)
pd3<- position_dodge(0.3)

#setwd don't forget


```


Figure 1 is methods diagram. 

Figure 2: Differentially abundant bacteria between C and HS flies. Data in figure 2 is from QIIME2 processing of 16S v1-v2 microbiome profiling. 
```{r}
#figure 2a
#plot for relative abundance
chs_nowolb <- readRDS("chs_nowolb_donors_obj_sp_SHARE.rds")


#remove thouse with fewer than 500 reads/fly
chs_red<- prune_samples(sample_sums(chs_nowolb)>=500, chs_nowolb)

#calculate relative abundance for visualization
chsred_donor<- chs_red%>%
    tax_glom(taxrank="Species") %>%
    transform_sample_counts(function(x) {x/sum(x)}) %>%
    psmelt() %>%
    #filters based on abundance at individual level
    filter(Abundance > 0.1) %>%
    arrange(Species)

#look at abundances 
chsred_donor %>%
  select(generation, Abundance, Species) %>%
  group_by(generation, Species) %>%
  filter(generation == "C donor fly") %>%
  summarise(avg=mean(Abundance)) %>%
  arrange(Species)

chsred_donor %>%
  select(generation, Sample, Abundance, Species) %>%
  filter(generation == "C donor fly" & Species == "Acetobacter persici") %>%
  summarise(n=n(), avg=mean(Abundance))

#create palette to highlight the differences between C Aceto and HS Aceto
donorpal<- c("#fb8072", "#80b1d3", "#8dd3c7", "#ffffb3", "#bebada", "#fdb462", "#b3de69", "#fccde5", "#d9d9d9","#bc80bd", "#ccebc5")


#individuals 
plot.donors.relabund.ind<- 
chsred_donor %>%
  #arbirtary order for visualization 
  mutate(artorder = case_when(generation == "C donor fly" ~ 1, 
                              generation == "HS donor fly" ~ 2)) %>%
  ggplot(., aes(x=fct_reorder(Sample, artorder), y=Abundance, fill=Species)) +
    geom_bar(stat="identity", position = "fill", width = 1,color="black", size=0.05) + 
    xlab("") + 
    ylab("Relative abundance") + 
    scale_fill_manual(values = donorpal) +
    #scale_fill_brewer(palette = "Set3") +
    scale_x_discrete(labels = c("C fly", "HS fly")) + 
    theme_classic() + 
    theme(panel.grid.major=element_blank(), panel.grid.minor=element_blank()) + 
    theme(panel.border = element_blank(), axis.line = element_line()) +
    #theme(axis.text.x = element_text(angle = 90, hjust = 1)) 
    theme(legend.position = "none") +
    theme(axis.text.x = element_blank(), axis.ticks.x = element_blank(), axis.line.x = element_blank())

#ggsave(plot.donors.relabund.ind, file="v9_plots/supp_donors_ind.pdf", height=6, width=7)

#calculate relative abundance by groups. save as df to speed up 
chsred_donor_rel<- chs_red %>%
    tax_glom(taxrank="Species") %>%
    transform_sample_counts(function(x) {x/sum(x)}) %>%
    psmelt()

#printing mean rel. abundance
chsred_donor_rel %>%
  group_by(generation, Species) %>%
  summarise(mean=mean(Abundance), ssd = sd(Abundance), se = sd(Abundance)/sqrt(n()), nsam =n()) %>%
  #arrange for high rel abund 
  arrange(desc(mean))

#figure 2b
#look at differences in ancom 
ancom.d.geno<- read.table("ancom_donor_flygeno.csv", sep=",", header=TRUE)

plot.ancom.d.geno<-
ggplot(ancom.d.geno, aes(x=clr, y=W)) + 
  geom_point(aes(fill=sig), shape=21, alpha=0.8, size=3) +
  scale_fill_manual(values = c("#99000d","#91bfdb","#737373"), 
                     labels = c("C fly", "HS fly", "Neutral")) + 
  theme_classic() +
  xlab("Mean centered log ratio") + 
  ylab("ANCOM W statistic") + 
  theme(legend.position="none") + 
  labs(fill="Significance")


#looking at the recipients 
chs_all <- readRDS("chs_nowolb_obj_sp.rds")

#have to filter for just recipients 
chs.recip <- chs_all %>%
  subset_samples(simple.gen == "C recipient" | simple.gen == "HS recipient") %>%
  prune_taxa(taxa_sums(.) > 0, .)

#need to clean up and normalize  
chs.recip.r <-
  chs.recip %>%
    subset_taxa(Kingdom == "Bacteria" & Family != "mitochondria" & Family != "chloroplast") %>%
    subset_taxa(Genus != "Wolbachia") %>%
    rarefy_even_depth(sample.size=500, replace=TRUE, trimOTUs=TRUE, rngseed=711) 

#ordinate 
chs.recip.bray <- ordinate(
  physeq = chs.recip.r, 
  method = "PCoA", 
  distance = "bray"
  )

#pcoa bray
#saving as dataframe to refine data viz
chs.recip.bray.pcoa<-
plot_ordination(
    physeq = chs.recip.r, 
    ordination = chs.recip.bray, 
    color="Sample_or_Control", 
    justDF = TRUE)

#axis 1 is 36.2% and axis 2 is 25.6%
#plot.chsrecip<-
ggplot(chs.recip.bray.pcoa, aes(x=Axis.1, y=Axis.2)) +
  geom_point(aes(color=treatment)) +
  facet_wrap(~food) + 
  theme_bw() +
  scale_color_manual(values=redt.microemph.pal) +
  theme(panel.grid = element_blank()) +
  xlab("Axis 1 (36.2%)") +
  ylab("Axis 2 (25.6%)") +
  theme(legend.position = "right") 

#ggsave(plot.chsrecip, file="chsrecip_pcoa_leg.pdf", height=2, width=4)

#want to look at differences in variance explained between diets
#filter by diet
chs.recip.c <- 
  chs.recip.r %>%
  subset_samples(food == "control diet") %>%
  prune_taxa(taxa_sums(.) > 0, .)

#make bray curtis data frame
chsrecip.c_bray <- phyloseq::distance(chs.recip.c, method = "bray")
#save c data frame
chsrecip.c.df <- data.frame(sample_data(chs.recip.c))
#run permanova 
adonis2(chsrecip.c_bray ~ genotype*microbiome, data = chsrecip.c.df)

chs.recip.hs <-
  chs.recip.r %>%
  subset_samples(food == "high sugar diet") %>%
  prune_taxa(taxa_sums(.) > 0, .)

#make bray curtis data frame
chsrecip.hs_bray <- phyloseq::distance(chs.recip.hs, method = "bray")
#save hs data frame 
chsrecip.hs.df <- data.frame(sample_data(chs.recip.hs))
#run permanova
adonis2(chsrecip.hs_bray ~ genotype*microbiome, data = chsrecip.hs.df)

#calculate the relative abundances
chs.recip.RA<- chs.recip.r %>%
    tax_glom(taxrank="Species") %>%
    transform_sample_counts(function(x) {x/sum(x)}) %>%
    psmelt() %>%
    #filters based on abundance at individual level
    filter(Abundance > 0.1) %>%
    arrange(Species)

chs.recip.RA %>%
  mutate(genomicro = paste0(genotype,"_", microbiome)) %>%
  ggplot(., aes(x=genomicro, y=Abundance, fill=Species)) + 
    geom_bar(stat = "identity", position="fill") +
    facet_wrap(~food)

#look at it this way 
#chs.recip.RA.all<- 
  chs.recip.r %>%
    tax_glom(taxrank="Species") %>%
    transform_sample_counts(function(x) {x/sum(x)}) %>%
    psmelt() %>% 
    mutate(genomicro = paste0(genotype,"+", microbiome)) %>%
    filter(Species == "Acetobacter indonesiensis 5H-1" | Species == "Acetobacter pasteurianus") %>%
      ggplot(., aes(x=food, y=Abundance)) + 
      geom_jitter(aes(color=Species), shape=21, size=0.3, alpha=0.2, width=0.1, stroke=1) +
      stat_summary(aes(group=Species, color=Species), fun=mean, geom="line", linewidth=1) +
      stat_summary(aes(group=Species, color=Species), fun.data=mean_se,
                   geom="pointrange", linewidth=1) +         
      facet_wrap(~microbiome) +
      theme_bw() + 
      theme(panel.grid = element_blank()) +
      scale_color_manual(values=c("#99000d","#91bfdb"))

chs.recip.RA.chsaceto <-
  chs.recip.r %>%
    tax_glom(taxrank="Species") %>%
    transform_sample_counts(function(x) {x/sum(x)}) %>%
    psmelt() %>%
    filter(Species == "Acetobacter indonesiensis 5H-1" | Species == "Acetobacter pasteurianus")

chs.recip.RA.chsaceto %>%
  filter(Species == "Acetobacter pasteurianus") %>%
  kruskal.test(data=., Abundance~food)


#want to track the abundance of C and HS Aceto over the transplant
#rarefy
chs.all.r <-
  chs_all %>%
    subset_samples(simple.gen == "C recipient" | simple.gen == "HS recipient" |
                   simple.gen == "C poop h20" | simple.gen == "HS poop h20") %>%
    subset_taxa(Kingdom == "Bacteria" & Family != "mitochondria" & Family != "chloroplast") %>%
    subset_taxa(Genus != "Wolbachia") %>%
    prune_taxa(taxa_sums(.) > 0, .) %>%
    rarefy_even_depth(sample.size=500, replace=TRUE, trimOTUs=TRUE, rngseed=711) 

#relative abundance
chs.all.r.ra <- 
 chs.all.r %>%
    tax_glom(taxrank="Species") %>%
    transform_sample_counts(function(x) {x/sum(x)}) %>%
    psmelt() %>% 
    mutate(genomicro = paste0(genotype,"+", microbiome)) %>%
    filter(Species == "Acetobacter indonesiensis 5H-1" | Species == "Acetobacter pasteurianus") 
 
#filter for hs poop h20 to rework the figure
chs.all.r.hspoop<- chs.all.r.ra %>% filter(simple.gen=="HS poop h20")

chs.all.r.hspoop$food <- "control diet"

chs.all.r.cpoop<- chs.all.r.ra %>% filter(simple.gen=="C poop h20")

chs.all.r.cpoop$food <- "high sugar diet"

rbind(chs.all.r.ra, chs.all.r.hspoop, chs.all.r.cpoop) %>%
      ggplot(., aes(x=simple.gen, y=Abundance)) +
      geom_jitter(aes(color=Species), shape=21, size=0.3, alpha=0.2, width=0.1, stroke=1) +
      stat_summary(aes(group=Species, color=Species), fun=mean, geom="line", linewidth=1) +
      stat_summary(aes(group=Species, color=Species), fun.data=mean_se,
                   geom="pointrange", linewidth=1) +
      facet_wrap(food~microbiome, scales="free_x") +
      theme_bw() +
      theme(panel.grid = element_blank()) +
      scale_color_manual(values=c("#99000d","#91bfdb"))

#supplemented dots of uric acid to see how acetobacters can break down
dots<- read.table("acetos_uricacid_210713b.csv", sep=",", header=TRUE)

#order microbe treatment
dots$microbe_f<- factor(dots$microbe, levels=c("No UA", "Sterile + UA", "C Aceto", "HS Aceto"))

#first just look at the addition between no ua and sterile to just show the increase worked
#plot.dots.comparesterile<-
dots %>%
  filter(microbe_f == "No UA" | microbe_f == "Sterile + UA") %>%
  ggplot(., aes(x=diet, y=um.ua, color=microbe_f)) +
  geom_boxplot(aes(color=microbe_f), outlier.shape = NA) +
  geom_point(aes(color=microbe_f), position=position_jitterdodge(jitter.width = 0)) +
  scale_color_manual(values=c("#bdbdbd", "#fee090")) +
  theme_classic() + 
  xlab("") + 
  ylab("Uric acid (umol)") +
  theme(legend.position = "right") + 
  theme(panel.grid = element_blank()) + 
  scale_x_discrete(labels = c("C diet", "HS diet")) + 
  theme(legend.title = element_blank())


#t test to compare either control or hs diet 
dots %>%
  filter(microbe_f == "No UA" | microbe_f == "Sterile + UA") %>%
  filter(diet == "high sugar") %>%
  t.test(data=., um.ua~microbe_f)
  
#paired plots
#plot.dots.ua <-
dots %>%
  filter(microbe_f != "No UA") %>%
  ggplot(., aes(x=microbe_f, y=um.ua, color=diet)) + 
  geom_boxplot(aes(color=diet), outlier.shape = NA) +
  geom_point(aes(fill = diet, color=microbe_f), position=position_jitterdodge(jitter.width = 0)) +
  #facet_wrap(~diet) + 
  scale_color_manual(values=c("#99000d", "#99000d", "#91bfdb", "#91bfdb", "#fee090")) +
  theme_classic() + 
  xlab("") + 
  ylab("Uric acid (umol)") +
  #theme(axis.text.x=element_text(angle=45, hjust=1)) + 
  theme(legend.position = "none") + 
  theme(panel.grid = element_blank()) +
  scale_x_discrete(labels = c("Sterile + UA", "C Aceto + UA", "HS Aceto + UA"))


#only No UA only has 4, the rest have 5 observations each 
dots %>%
  group_by(diet, microbe_f) %>%
  summarise(n=n())

#calculating average to find percent decrease on hs diet 
dots %>%
  filter(diet == "high sugar" & microbe_f != "No UA" & microbe_f!="Sterile + UA") %>%
  group_by(microbe_f) %>%
  summarise(avg=mean(um.ua)) 

#c aceto is 64.7 and hs aceto is 30.3
(64.7-30.3)/64.7*100

#remove the no ua treatment from stats because more a technical artefect
dots.allua<- filter(dots, microbe != "No UA")

dots.m2<- lm(data=dots.allua, um.ua~microbe_f*diet)
car::Anova(dots.m2, type="III")

#check assumptions
#residuals normally distributed w=0.977, p=0.7694
hist(resid(dots.m2))
shapiro.test(resid(dots.m2))

#check qqplot. resids look good 
ggpubr::ggqqplot(dots.m2$residuals)
#check fitted vs residuals. 3 outliers, but not super weird
plot(dots.m2, 1)
#homogeneity of variance is OK 
car::leveneTest(data=dots.allua, um.ua~microbe_f*diet)

#put it all together
#make a new figure 2 
#plot.donors.relabund.ind, plot.ancom.d.geno
#use ggpubr to arrange
fig.2<- ggarrange(plot.donors.relabund.ind, 
          ggarrange(plot.ancom.d.geno, plot.dots.ua, ncol=2, widths =c(1,1), labels=c("B","C")), 
          nrow=2, labels="A")

ggsave(fig.2, file="~/Desktop/fig2abc.pdf", height=5, width=10)

```


figure 3 is the phenotypic effects from Acetobacter transplant for fecundity

```{r}
ovi.data<- read.table("2536_aceto_ovidata.csv", sep=",", header=TRUE)

#reorder treatment levels
ovi.data$fly.treat_f<- factor(ovi.data$fly.treat, levels = c("C0C", "CCC", "CHSC", 
                                                              "HS0C", "HSCC", "HSHSC", 
                                                              "C0HS", "CCHS", "CHSHS", 
                                                              "HS0HS", "HSCHS", "HSHSHS"))

ovi.data$fly.micro_f<- factor(ovi.data$fly.micro, levels = c("sterile", "control", "high sugar"))

#ovitest was done on two different media, but there was no effect
ggplot(ovi.data, aes(x=ovi.plate.treat, y=egg.no)) + 
    geom_boxplot()

wilcox.test(data=ovi.data, egg.no~ovi.plate.treat)

#get sample sizes
ovi.data %>%
  group_by(fly.geno, fly.micro, fly.food) %>%
  summarise(n=n())


#visualize histogram of egg counts
#lots of zeros 
ggplot(ovi.data, aes(x=egg.no)) + 
    geom_histogram(binwidth=1) +
    theme_classic() 

#visualize by histogram
ggplot(ovi.data, aes(x=egg.no, fill=fly.micro_f)) + 
  facet_wrap(fly.food~fly.geno) +
  geom_histogram(aes(color=fly.micro_f), alpha = 0.6, binwidth = 2, position="dodge") + 
  scale_color_manual(values=c("#fee090", "#99000d", "#91bfdb"), labels = c("sterile", "C Aceto", "HS Aceto")) + 
  scale_fill_manual(values=c("#fee090", "#99000d", "#91bfdb"), labels = c("sterile", "C Aceto", "HS Aceto")) + 
  theme_bw() + 
  theme(panel.grid.major=element_blank(), panel.grid.minor=element_blank()) + 
  xlab("Egg count") + 
  ylab("Frequency") + 
  theme(legend.position = "bottom")


#model to include zero dispersion through poisson distribution
egg.m1<- glmmTMB(data=ovi.data, egg.no ~ (fly.geno * fly.micro_f * fly.food ) + (1|plate.no), zi=~1, family=poisson)
#update constrasts
egg.m1<- update(egg.m1, contrasts=list(fly.geno="contr.sum", fly.micro_f="contr.sum", fly.food="contr.sum"))

#check dispersions
#some dispersion but 1.34
testDispersion(egg.m1)
#zero inflation is good 
testZeroInflation(egg.m1)

summary(egg.m1)
Anova(egg.m1, type="III")

#check without interactions
egg.m2<- glmmTMB(data=ovi.data, egg.no ~ (fly.geno + fly.micro_f + fly.food ) + (1|plate.no), zi=~1, family=poisson)
egg.m2<- update(egg.m2, contrasts=list(fly.geno="contr.sum", fly.micro_f="contr.sum", fly.food="contr.sum"))

#check dispersions
#no dispersion
testDispersion(egg.m2)
#zero inflation is good 
testZeroInflation(egg.m2)

#stat summary
summary(egg.m2)
Anova(egg.m2, type="III")

#compare models
anova(egg.m1, egg.m2)

#visual interactions through least square means 
emmip(egg.m1, fly.micro_f~fly.food|fly.geno)

#save to customize visualization. want diet on a axis 
m1.emm<- emmip(egg.m1, fly.micro_f~fly.food|fly.geno, CIs = TRUE, plotit=FALSE)

#weird issue with too many geom_s; rename variables 
levels(m1.emm$fly.micro_f)[levels(m1.emm$fly.micro_f) == "control"] <- "C Aceto"
levels(m1.emm$fly.micro_f)[levels(m1.emm$fly.micro_f) == "high sugar"] <- "HS Aceto"
levels(m1.emm$fly.food)[levels(m1.emm$fly.food) == "control diet"] <- "C diet"
levels(m1.emm$fly.food)[levels(m1.emm$fly.food) == "high sugar diet"] <- "HS diet"
levels(m1.emm$fly.geno)[levels(m1.emm$fly.geno) == "control fly"] <- "C fly"
levels(m1.emm$fly.geno)[levels(m1.emm$fly.geno) == "high sugar fly"] <- "HS fly"

#interaction plot 
ggplot(m1.emm, aes(x=fly.food, y=yvar, group=interaction(fly.geno, fly.micro_f))) + 
  geom_point(aes(color=fly.micro_f), position=pd, size=2.5) + 
  geom_line(aes(color=fly.micro_f), position=pd) +
  facet_wrap(~fly.geno) + 
  scale_color_manual(values = c("#fee090", "#99000d", "#91bfdb")) + 
  geom_errorbar(aes(ymin=yvar-SE, ymax=yvar+SE, color=fly.micro_f, width=0.1), position = pd) +
  theme_bw() + 
  xlab("") + 
  ylab("Linear prediction") +
  ylab("Least square means for fecundity") + 
  theme(legend.title=element_blank()) + 
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  ylim(0,3) 


#rework tbe interaction plot to go by diet 
emmip(egg.m1, fly.micro_f~fly.geno|fly.food)

#save to customize visualization. want diet on a axis 
m1.emmB<- emmip(egg.m1, fly.micro_f~fly.geno|fly.food, CIs = TRUE, plotit=FALSE)

plot.fecundity.revise<-
ggplot(m1.emmB, aes(x=fly.geno, y=yvar, group=interaction(fly.geno, fly.micro_f))) + 
  geom_point(aes(color=fly.micro_f), position=pd, size=2.5) + 
  geom_line(aes(color=fly.micro_f, group=fly.micro_f), position=pd) +
  facet_wrap(~fly.food) + 
  scale_color_manual(values = c("#fee090", "#99000d", "#91bfdb")) + 
  geom_errorbar(aes(ymin=yvar-SE, ymax=yvar+SE, color=fly.micro_f, width=0.1), position = pd) +
  theme_bw() + 
  xlab("") + 
  ylab("Linear prediction") +
  ylab("Least square means for fecundity") + 
  scale_x_discrete(labels = c("control fly" = "C fly", "high sugar fly" = "HS fly")) +
  theme(legend.title=element_blank()) + 
  theme(panel.grid = element_blank()) +
  theme(legend.position = "none") +
  ylim(0,3) 

ggsave(plot.fecundity.revise, file="~/Desktop/plot_3b_revise.pdf", height=3, width=3)

#print counts by different groupings
#for diet differences
ovi.data %>%
  group_by(fly.food) %>%
  summarise(mean=mean(egg.no), sd = sd(egg.no), se = sd/sqrt(n()))

#for differences between sterile and non sterile
ovi.data %>%
  mutate(type = case_when(fly.micro == "sterile" ~ "sterile", 
                          fly.micro != "sterile" ~ "microbes")) %>%
  group_by(fly.food, type) %>%
  summarise(mean=mean(egg.no), sd = sd(egg.no), se = sd/sqrt(n()))

#for 3way interactions 
ovi.data %>%
  group_by(fly.geno,fly.micro_f, fly.food) %>%
  summarise(N=n(), mean=mean(egg.no), sd = sd(egg.no), se = sd/sqrt(n()))


```






